### ISSUE TITLE
IPFS (`nerdctl run /ipfs/QmfTVLXMG9TH7X523NytcXj35XtEDx4wgNWYepVumqpJZV`)

### ISSUE THREAD

**Issue by AkihiroSuda** [containerd/nerdctl#465]
Stargz Snapshotter is going to support IPFS soon (https://github.com/containerd/stargz-snapshotter/pull/496).
We have to decide how CLI will look like for nerdctl + IPFS.

## Pull & Run
```console
$ nerdctl pull /ipfs/QmfTVLXMG9TH7X523NytcXj35XtEDx4wgNWYepVumqpJZV
$ nerdctl pull /ipns/example.com/foo
```

ref: https://docs.ipfs.io/concepts/dnslink/

~I think we can just assume that the image ref string is an IPFS multihash when the ref string starts with `Qm`~
(**EDIT**: Only CID v0 always starts with `Qm`. CID v1 does not start with `Qm` https://docs.ipfs.io/concepts/content-addressing/#identifier-formats)

We should propose the same image ref syntax to Kubernetes too.

## Push
???

## imgcrypt
???

@ktock

**Comment by ktock** [comment]
@AkihiroSuda 

I'll make a draft PR for push/pull in days.

How can we support IPFS in `nerdctl build`? (i.e. pulling base images from IPFS)
Though we need to modify BulidKit's resolver logic eventually, a possible workaround might be to fix nerdctl to pre-pull base images from IPFS before the build and import them to BuildKit via local cache importer.
(But how can nerdctl know base images needed for a build?)

**Comment by AkihiroSuda** [comment]
> @AkihiroSuda
> 
> I'll make a draft PR for push/pull in days.


thx :+1:

> How can we support IPFS in `nerdctl build`? (i.e. pulling base images from IPFS) Though we need to modify BulidKit's resolver logic eventually, a possible workaround might be to fix nerdctl to pre-pull base images from IPFS before the build and import them to BuildKit via local cache importer. (But how can nerdctl know base images needed for a build?)

Can we just have a local OCI registry that proxies OCI Dist Spec requests to IPFS?

**Comment by ktock** [comment]
> Can we just have a local OCI registry that proxies OCI Dist Spec requests to IPFS?

Maybe we can use IPDR (https://github.com/ipdr/ipdr) here.

**Comment by AkihiroSuda** [comment]
Implemented in #505 #525
Kudos @ktock
